pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,355 @@
+	listView            *view.ListView
+	ctx                 context.Context
+	taskToChannelMap    map[TaskDetails]chan Result
+	shouldStopListening bool
+}
+
+func (t *TaskViewManagerImpl) ResetVirtualCenter(virtualCenter *cnsvsphere.VirtualCenter) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.virtualCenter = virtualCenter
+}
+
+func getListViewWaitFilter(listView *view.ListView) *property.WaitFilter {
+	ts := types.TraversalSpec{
+		Type: "ListView",
+		Path: "view",
+		Skip: types.NewBool(false),
+	}
+	filter := new(property.WaitFilter)
+
+	filter.Add(listView.Reference(), "Task", []string{infoPropertyName}, &ts)
+	reportMissingObjectsInResults := true
+	filter.Spec.ReportMissingObjectsInResults = &reportMissingObjectsInResults
+	return filter
+}
+
+// listenToTaskUpdates is a long-running goroutine
+// that uses a property collector to listen for task updates
+// CSI ops add CNS tasks to listview and wait for a response from CNS
+// when update(s) are received by the property collector,
+// it spawns a new goroutine to process each task update and return the result to the caller
+func (t *TaskViewManagerImpl) listenToTaskUpdates(ctx context.Context) {
+	log := logger.GetLogger(ctx)
+	var filter *property.WaitFilter
+	recreateView := false
+	for {
+
+		if t.ctx.Err() == context.Canceled {
+			log.Info("listener canceled")
+			break
+		}
+
+		if t.getListView() == nil || recreateView {
+			log.Infof("listenToTaskUpdates -> recreateView -> createTasksListView")
+			listView, err := createTasksListView(ctx, t.virtualCenter.Client.Client, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				time.Sleep(waitForUpdatesRetry)
+				continue
+			}
+			t.setListView(t.ctx, listView)
+		}
+
+		if filter == nil || recreateView {
+			filter = getListViewWaitFilter(t.getListView())
+			recreateView = false
+		}
+		log.Infof("Starting listening for updates...")
+		pc := property.DefaultCollector(t.virtualCenter.Client.Client)
+		err := property.WaitForUpdates(t.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Infof("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Infof("Got update for object %v properties %v", update.Obj, prop)
+					if prop.Name == infoPropertyName {
+						go t.handleTaskUpdate(ctx, prop, false)
+					}
+				}
+			}
+
+			return t.shouldStopListening
+		})
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v, vimClient: %v", err, t.virtualCenter.Client.Client.UserAgent)
+			if soap.IsSoapFault(err) {
+				switch soap.ToSoapFault(err).VimFault().(type) {
+				case *types.ManagedObjectNotFound, types.ManagedObjectNotFound:
+					recreateView = true
+				case *types.NotAuthenticated, types.NotAuthenticated:
+					log.Info("observed NotAuthenticated error. should recreate view\n")
+					// todo: do we need to recreate view here?
+					recreateView = true
```

Comment:
Why recreateView?

Reply:
ListView is created by using a vimClient reference. 

`listView, err := view.NewManager(vimClient).CreateListView(ctx, tasks)`

When the old vimClient reference is no longer valid, we need to recreate the view otherwise we keep getting NotAuthenticated errors when adding/removing items from the listView. I removed the todo comment.